include/psa/crypto_se_driver.h

```diff
@@ -677,6 +677,49 @@ typedef struct {
 } psa_drv_se_aead_t;
 /**@}*/
 
+/** \defgroup se_slot_usage Secure element slot usage mask
+ */
+/**@{*/
+
+/** The type of slot usage data for a driver.
+ *
+ * A driver can use this data to keep track of which slot numbers are in use
+ * and which are available for new keys. The core stores this data to
+ * internal persistent storage.
+ *
+ * The representation of this type is opaque. To access it,
+ * the driver can call psa_drv_cb_find_free_slot().
+ */
+typedef struct psa_drv_se_slot_usage_s psa_drv_se_slot_usage_t;
+
+/** Callback function to find a free slot within a range.
+ *
+ * The PSA Crypto core provides this function to access the opaque
+ * data type that represents the slot usage.
+ *
+ * \param[in] slot_usage    The opaque data structure containing the
+ *                          driver's slot usage table.
+ * \param from              The start of the range.
+ *                          It is included in the search.
+ * \param before            The end of the range.
+ *                          It is not included in the search.
+ * \param[out] found        On success, a free slot number such that
+ *                          `from <= found < before`.
```

Comment:
What is the reason behind such rule? It will be hard to remember, since it's inconsistent even within this one function call.

Reply:
I don't understand. Why is it so strange that the slot is within an interval, since that's the whole point of the function? What inconsistency are you talking about?

Reply:
I'm talking about having the beginning of the range inclusive, but the end - not.

Reply:
That's the usual way to do ranges, so that end - start = length. What would you expect?

Reply:
I would expect having the same rules for both sides of the range. I'm not acquainted with other crypto APIs though, so I won't press on changing it.

Reply:
It's not about crypto APIs, it's about array slices. Most array slicing APIs take (start_inclusive, number_of_items) or (start_inclusive, end_exclusive) as parameters, with end_exclusive = start_inclusive + number_of_items. If you use start_inclusive and end_inclusive, a lot of operations need a +1 or -1, which is awkward and error-prone.

Reply:
But maybe this would look more familiar with a length parameter instead of an end parameter. I'm considering changing it.

Reply:
Moot because this interface has been completely redesigned.